lib/vsc/modules/cache.py

```diff
@@ -0,0 +1,293 @@
+    # only one element, and it is a single-element list
+    return get_lua_via_json(LMOD_CONFIG, ['scDescriptT'])[0][0]
+
+
+def get_lmod_cache(cachefile):
+    """Return Lmod lua cache as list of modulepaths and spider data"""
+    return get_lua_via_json(cachefile, ['mpathMapT', 'spiderT'])
+
+
+def cluster_map(mpathMapT):
+    """Return cluster -> cluster module and modulepath -> [cluster1, ...] mappings"""
+    # map cluster to cluster module
+    clustermap = {}
+    # map modulepath to list of clusters
+    modulepathmap = {}
+    for mpath, data in mpathMapT.items():
+        for clmod in data.keys():
+            if not clmod.startswith('cluster/'):
+                continue
+            cluster = clmod.split('/')[1].lstrip('.')
+            tmpclmod = clustermap.setdefault(cluster, clmod)
+            if tmpclmod != clmod:
+                LOGGER.raiseException("Found 2 different cluster modules %s and %s for same cluster %s" %
+                                      (tmpclmod, clmod, cluster))
+            mpclusters = modulepathmap.setdefault(mpath, [])
+            if cluster not in mpclusters:
+                mpclusters.append(cluster)
+
+    LOGGER.debug("Generated clustermap %s", clustermap)
+    LOGGER.debug("Generated modulepathmap %s", modulepathmap)
+    return clustermap, modulepathmap
+
+
+def sort_modulepaths(spiderT, mpmap):
+    """Return a sorted list of modulepaths"""
+    modulepaths = []
+    for mpath in spiderT.keys():
+        if not mpath.startswith("/"):
+            LOGGER.debug("Skipping spiderT key %s", mpath)
+            continue
+
+        if mpath in mpmap:
+            modulepaths.append(mpath)
+        else:
+            LOGGER.debug("Skipping modulepath %s not in modulepath map %s", mpath, mpmap)
+
+    modulepaths.sort()
+    LOGGER.debug("Found pre-sorted modulepaths %s", modulepaths)
+
+    # sort them
+    #   very trivial sort based on EXTRA_MODULEPATHS from CLUSTER_DATA
+    #   we do not assume complicated nesting
+    #   just search all EXTRA_MODULEPATHS last
+    #      they are listed in prepend order, but default/most sensible one is prepended last
+    #      so they are reversed: the first extra has to be moved as far as possible
+    for extras in [x.get('EXTRA_MODULEPATHS', [])[::-1] for x in CLUSTER_DATA.values()]:
+        for extra in extras:
+            # move to the end
+            LOGGER.debug("Moving EXTRA_MODULEPATH %s to the end", extra)
+            try:
+                modulepaths.remove(extra)
+                modulepaths.append(extra)
+            except ValueError as e:
+                LOGGER.error("Did not find EXTRA_MODULEPATH %s in modulespaths %s: %s", extra, modulepaths, e)
+
+    LOGGER.debug("Sorted modulepaths %s", modulepaths)
+    return modulepaths
+
+
+def sort_recent_versions(versions):
+    """Sort versions using LooseVersion, most recent first"""
+    looseversions = map(LooseVersion, versions)
+    return [x.vstring for x in sorted(looseversions)][::-1]
```

Comment:
Isn't this basically a complicated way of doing `sorted(versions, key=LooseVersion, reverse=True)`?

Reply:
ah, i didn't know `key=` worked like that.